weechat: allow specifying dicts for spellchecking

expipiplus1 opened the pull request:
Use as so:

```nix 
weechat = super.weechat.override {
 weechatEnchantHunspellDicts = [pkgs.hunspellDicts.en-us];
}
```

```nix 
weechat = super.weechat.override {
 weechatAspellDicts = [pkgs.aspellDicts.en];
}
```

expipiplus1 committed:
weechat: Add plugin to config only if it is used

Previously it wasn't possible to build weechat without python support as
the `weechat.python` attribute was used in the config'

expipiplus1 committed:
weechat: Put python interpreter in path

fixes #11646

expipiplus1 committed:
weechat: Remove unused parameter to default.nix

expipiplus1 committed:
weechat: Enable Enchant spell checking

expipiplus1 committed:
weechat: allow specifying dicts for spellchecking

Use as so:

```nix
weechat = super.weechat.override {
  weechatEnchantHunspellDicts = [pkgs.hunspellDicts.en-us];
}
```

```nix
weechat = super.weechat.override {
  weechatAspellDicts = [pkgs.aspellDicts.en];
}
```

FRidh commented:
Related PR #34856

FRidh commented on pkgs/applications/networking/irc/weechat/default.nix:
```diff
@@ -26,6 +34,16 @@ let
   ];
   enabledPlugins = builtins.filter (p: p.enabled) plugins;
 
+  aspellDictEnv = buildEnv{
```
Can you use `aspellWithDicts`?

expipiplus1 replied:
Good idea, I don't think the same can be done for the enchant/hunspell pair without a bit more plumbing though.

expipiplus1 replied:
@FRidh done

FRidh replied:
> Why not just install them globally with an enchant module, instead of overriding each package separately?

One may want to be explicit about what dictionaries are used.

But I agree with you; in practice one likely just wants what is provided by the system (`XDG_DATA_DIRS`).

expipiplus1 replied:
That didn't seem to be working in my case, which motivated this PR. I think there's value in having this all self contained though, indeed I'm currently running weechat on a server with a user without XDG_DATA_DIRS.

I do agree that patching each user of `enchant` isn't a perfect solution though. Installing them globally doesn't quite fit the nix philosophy though. A much better change would be to patch aspell so that not only the binary can find the dictionaries in `aspellWithDicts` but the library can too (and similarly for hunspell).

expipiplus1 replied:
One small downside of that is that spelling dictionaries seem to be a fairly fragile part of nixpkgs, and meddling with aspell/hunspell internals might have to be done carefully.

expipiplus1 committed:
weechat: use aspellWithDicts to get dict paths

Use like this now:

```nix
weechat = super.weechat.override {
  aspell = pkgs.aspellWithDicts (ps: [ps.en]);
};
```

expipiplus1 committed:
weechat: remove weechat prefix from parameter name

jtojnar commented:
I can imagine someone using a specific program only for communication in a single language and not needing other dictionaries installed system-wide but I think this use case is quite rare and not really worth the added complexity of package expressions. This is exactly what environment is for.

:-1:

bemeurer commented:
This is great!

bdesham commented:
In case it’s useful to anyone else, here’s a way to get this same kind of customizability for WeeChat and Aspell without patching nixpkgs. Add a section like the following to the `packageOverrides` in your `config.nix`:

```nix
{ pkgs }:

{
  packageOverrides = pkgs: rec {

    weechat = let
      aspell = pkgs.aspellWithDicts (dicts:
        with dicts; [ en en-computers en-science ]
      );
    in (pkgs.weechat.overrideAttrs (oldAttrs: {
      buildInputs = (oldAttrs.buildInputs or []) ++ [ aspell ];
    })).override {
      configure = { availablePlugins, ... }: {
        init = ''
          /set env ASPELL_CONF "dict-dir ${aspell}/lib/aspell"
        '';
      };
    };

  };
}
```

This example selects the `en`, `en-computers`, and `en-science` dictionaries from Aspell.

expipiplus1 commented:
@bdesham Looks great, I'm using this now

expipiplus1 commented:
Closing as @bdesham's solution is neater.
